src/solver/cartesianprod/forward_cartesianprod.cpp

```diff
@@ -0,0 +1,162 @@
+    {
+        return false;
+    }
+    if(!problem.IsAllPacked())
+    {
+        return false;
+    }
+    return true;
+}
+
+ConvSolution
+CartesianProdForward::GetSolution(const ExecutionContext& context,
+                                  const miopen::cartesianprod::FwdProblemDescription& problem) const
+{
+    std::ignore       = context;
+    auto output_dtype = miopen::GetDataType(problem.GetOutputDesc().GetType());
+    auto dtype        = problem.GetOutputDesc().GetType();
+    auto inputCount   = problem.GetInputCount();
+
+    auto build_params =
+        KernelBuildParameters{{"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+                              {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+                              {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+                              {"TILE_SIZE", TILE_SIZE},
+                              {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype}};
+
+    if(inputCount == 1)
+    {
+        return ConvSolution{miopenStatusNotImplemented};
+    }
+    auto output_dims = problem.GetOutputDesc().GetLengths();
+    auto result      = ConvSolution{miopenStatusSuccess};
+    result.construction_params.push_back(make_hip_kernel({LOCAL_SIZE_FWD},
+                                                         {output_dims[0]},
+                                                         "MIOpenCartesianProd.cpp",
+                                                         "CartesianProdForward",
+                                                         build_params));
+    result.construction_params.push_back(make_hip_kernel({TILE_SIZE, TILE_SIZE},
+                                                         {output_dims[0], output_dims[1]},
+                                                         "MIOpenCartesianProd.cpp",
+                                                         "CartesianProdTranspose",
+                                                         build_params));
+
+    result.invoker_factory = [inputCount](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            size_t stride = 1;
+            HipEventPtr start, stop;
+            bool profiling = handle_.IsProfilingEnabled();
+            if(profiling)
+            {
+                handle_.EnableProfiling(false);
+                hipStreamSynchronize(handle_.GetStream());
+                start = miopen::make_hip_event();
+                stop  = miopen::make_hip_event();
+                hipEventRecord(start.get(), handle_.GetStream());
+            }
+
+            decltype(auto) kernel = handle_.Run(kernels[0]);
+            decltype(auto) params = raw_params.CastTo<miopen::cartesianprod::FwdInvokeParams>();
+            auto output_tv        = get_inner_expanded_tv<2>(deref(params.outputDesc));
+
+            par_for(inputCount, [&](auto i) {
```

Comment:
Parallelism is meaning less, because all kernels are scheduled on the same stream.

Reply:
resolved